### PR TITLE
[FIR] add CLI flag to run extended checkers and add messages for extended warnings

### PR DESCRIPTION
**Pull request by vldF**

### CODE REVIEW

File: compiler/cli/src/org/jetbrains/kotlin/cli/jvm/compiler/KotlinToJVMBytecodeCompiler.kt [hunk 1]
```diff
@@ -197,8 +198,10 @@ object KotlinToJVMBytecodeCompiler {
         }
 
         val projectConfiguration = environment.configuration
+        val extendedAnalysisMode = projectConfiguration.getBoolean(CommonConfigurationKeys.USE_FIR_EXTENDED_CHECKERS)
```
Comment (mglukhikh) [inline compiler/cli/src/org/jetbrains/kotlin/cli/jvm/compiler/KotlinToJVMBytecodeCompiler.kt:201]: It's better to calculate it under the following `if`

**Comment by vldF** [comment]
RR branch: https://github.com/JetBrains/kotlin/tree/rr/vldf/extended-checkers-cli

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDefaultErrorMessages.kt [hunk 2]
```diff
@@ -199,17 +210,17 @@ class FirDefaultErrorMessages : DefaultErrorMessages.Extension {
...
-//            map.put(REDUNDANT_RETURN_UNIT_TYPE, ...) // &
-//            map.put(REDUNDANT_EXPLICIT_TYPE, ...) // &
-//            map.put(REDUNDANT_SINGLE_EXPRESSION_STRING_TEMPLATE, ...) // &
-//            map.put(CAN_BE_VAL, ...) // &
-//            map.put(CAN_BE_REPLACED_WITH_OPERATOR_ASSIGNMENT, ...) // &
-//            map.put(REDUNDANT_CALL_OF_CONVERSION_METHOD, ...) // &
-//            map.put(ARRAY_EQUALITY_OPERATOR_CAN_BE_REPLACED_WITH_EQUALS, ...) // &
-//            map.put(EMPTY_RANGE, ...) // &
-//            map.put(REDUNDANT_SETTER_PARAMETER_TYPE, ...) // &
+            map.put(REDUNDANT_VISIBILITY_MODIFIER, "redundant visibility modifier")
```
Comment (mglukhikh) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDefaultErrorMessages.kt:213]: It's better (and more consistent) to start messages from a capital letter

File: compiler/testData/cli/js/jsExtraHelp.out [hunk 3]
```diff
@@ -70,6 +70,7 @@ where advanced options include:
   -Xskip-prerelease-check    Allow to load pre-release classes
   -Xuse-experimental=<fq.name> Enable, but don't propagate usages of experimental API for marker annotation with the given fully qualified name
   -Xuse-fir                  Compile using Front-end IR. Warning: this feature is far from being production-ready
+  -Xuse-fir-extended-checkers Use fir extended checkers. Warning: this feature is far from being production-ready
```
Comment (mglukhikh) [inline compiler/testData/cli/js/jsExtraHelp.out:73]: I'd write something more clear, like "Use extended analysis mode based on Front-end IR. Warning: ..."

File: compiler/cli/src/org/jetbrains/kotlin/cli/jvm/compiler/KotlinToJVMBytecodeCompiler.kt [hunk 4]
```diff
@@ -384,6 +394,8 @@ object KotlinToJVMBytecodeCompiler {
                 environment.messageCollector
             )
 
+            if (extendedAnalysisMode) return false
```
[the remark this answers is not in the document]
Comment (vldF) [reply compiler/cli/src/org/jetbrains/kotlin/cli/jvm/compiler/KotlinToJVMBytecodeCompiler.kt:397]: The reason for this condition in my idea of extended checkers. By my opinion, they must be executed without binary compiling. When we run extended checkers, we would like to get suggestions to the code, not compiled binaries. Is this opinion right?

[the remark this answers is not in the document]
Comment (zarechenskiy) [reply compiler/cli/src/org/jetbrains/kotlin/cli/jvm/compiler/KotlinToJVMBytecodeCompiler.kt:397]: Aha, now I see, thanks. It's discussable but I believe we can merge your changes now

[the remark this answers is not in the document]
Comment (vldF) [reply compiler/cli/src/org/jetbrains/kotlin/cli/jvm/compiler/KotlinToJVMBytecodeCompiler.kt:397]: Ok, we can discuss it on the next meeting

**Comment by mglukhikh** [comment]
Merged manually 👍

**Pull request closed without merging**